Run frr service as frr_t and add new bfd_multi port

Koncpa opened the pull request:
Confined frr service as frr_t and create policy files and rules for frr service.
FRRouting (FRR) is an IP routing protocol which
includes protocol daemons for BGP, IS-IS, LDP, OSPF, PIM, and RIP.
Add new port called bfd_multi for bfdd daemon, which is use in frr service.

WOnder93 commented on policy/modules/contrib/frr.if:
```diff
@@ -0,0 +1,160 @@
+
+    tunable_policy(`deny_ptrace',`',`
+        allow $1 frr_t:process ptrace;
+    ')
```
There is some whitespace inconsistency here.

Koncpa replied:
This part of policy I generate via sepolicy generate, so It's possible that here have appeared some whitespace inconsistency. Thanks, I fix this.

WOnder93 commented on policy/modules/contrib/frr.te:
```diff
@@ -0,0 +1,81 @@
+#
+# frr local policy
+#
+allow frr_t self:capability { dac_override kill net_bind_service net_raw setgid setuid };
```
Did you investigate why frr needs the `dac_override` capability? Sometimes it's caused by wrong permissions being set on files.

Koncpa replied:
Yes, I investigate this issue, the frr need delete pid file **/var/run/frr/bgpd.pid** so need bypass ownership of the **rm** to delete this file. If you want I can send you full audit message.

WOnder93 replied:
Looking at the daemons the service starts, it might indeed be unavoidable if the file is deleted by `watchfrr` (which runs as root) instead of `bgpd` itself. Please paste here the full audit record so that my curiosity is satisfied ;)

Koncpa replied:
Here you have :)
`type=PROCTITLE msg=audit(01/21/2021 09:15:43.987:358) : proctitle=rm -f /var/run/frr/bgpd.pid
type=PATH msg=audit(01/21/2021 09:15:43.987:358) : item=1 name=/var/run/frr/bgpd.pid inode=33039 dev=00:18 mode=file,644 ouid=frr ogid=frr rdev=00:00 obj=system_u:object_r:frr_var_run_t:s0 nametype=DELETE cap_fp=none cap_fi=none cap_fe=0 cap_fver=0 cap_frootid=0
type=PATH msg=audit(01/21/2021 09:15:43.987:358) : item=0 name=/var/run/frr/ inode=27604 dev=00:18 mode=dir,755 ouid=frr ogid=frr rdev=00:00 obj=system_u:object_r:frr_var_run_t:s0 nametype=PARENT cap_fp=none cap_fi=none cap_fe=0 cap_fver=0 cap_frootid=0
type=CWD msg=audit(01/21/2021 09:15:43.987:358) : cwd=/
type=SYSCALL msg=audit(01/21/2021 09:15:43.987:358) : arch=x86_64 syscall=unlinkat success=yes exit=0 a0=0xffffff9c a1=0x55614f7d0640 a2=0x0 a3=0x0 items=2 ppid=4755 pid=4780 auid=unset uid=root gid=root euid=root suid=root fsuid=root egid=root sgid=root fsgid=root tty=(none) ses=unset comm=rm exe=/usr/bin/rm subj=system_u:system_r:frr_t:s0 key=(null)
type=AVC msg=audit(01/21/2021 09:15:43.987:358) : avc:  denied  { dac_override } for  pid=4780 comm=rm capability=dac_override  scontext=system_u:system_r:frr_t:s0 tcontext=system_u:system_r:frr_t:s0 tclass=capability permissive=1
`

WOnder93 replied:
Yes, so it's a root process (presumably watchfrr) removing a pid file owned by user frr... in that case the dac_override is justified.

WOnder93 commented on policy/modules/contrib/frr.te:
```diff
@@ -0,0 +1,81 @@
+
+domain_use_interactive_fds(frr_t)
+
+files_manage_generic_locks(frr_t)
```
What exact lock file is frr trying to access? If it's its own lock file, then perhaps it should be labeled as `frr_lock_t`, avoiding the need for this generic rule?

Koncpa replied:
Frr trying access to lock files like **/var/lock/susbys/bgpd**. This lock files are related with protocols, which frr use. I think, that we are not able to label dir /varl/lock/subsys as frr_lock_t, because dir is not use only for frr lock files, but It's possible to label each lock file which use frr or what do you think?

Koncpa replied:
I tried to label each lock file and test case with these changes was succesfull without any fail and AVC's, so I update this PR.

WOnder93 replied:
The `files_manage_generic_locks(frr_t)` is still there in the last version - can you try with it removed?

Koncpa replied:
Oh sorry, I forgot to delete it in PR version. Now It should be everything updated.

WOnder93 reviewed (approved):
Looks good to my untrained eye now, although @zpytela might have further comments.

Koncpa commented:
Add another permissions for frr policy, which were discover during testing a common selinux policy tests. The build test will fail, because in policy has added interface, which isn't yet in policy, but PR has already made: https://github.com/fedora-selinux/selinux-policy/pull/556

WOnder93 commented:
> The build test will fail, because in policy has added interface, which isn't yet in policy, but PR has already made: #556

In that case why not just include the commit in this PR instead of opening a separate one?

zpytela commented on policy/modules.conf:
```diff
@@ -3064,3 +3064,11 @@ rrdcached = module
+# Layer: contrib
+# Module: frr
+#
+# frr
```
Abbreviations usually deserve a few explanatory words.

Koncpa replied:
Fixed.

zpytela commented on policy/modules/contrib/frr.fc:
```diff
@@ -0,0 +1,25 @@
+/etc/frr(/.*)?                  gen_context(system_u:object_r:frr_conf_t,s0)
+
+/var/log/frr(/.*)?              gen_context(system_u:object_r:frr_log_t,s0)
+/var/tmp/frr(/.*)?             gen_context(system_u:object_r:frr_tmp_t,s0)
```
Misaligned line.

Koncpa replied:
Fixed.

zpytela commented on policy/modules/contrib/frr.fc:
```diff
@@ -0,0 +1,25 @@
+/var/lock/subsys/ripd           gen_context(system_u:object_r:frr_lock_t,s0)
+/var/lock/subsys/ripngd         gen_context(system_u:object_r:frr_lock_t,s0)
+/var/lock/subsys/staticd        gen_context(system_u:object_r:frr_lock_t,s0)
+/var/lock/subsys/zebra          gen_context(system_u:object_r:frr_lock_t,s0)
```
Are all of the subsys locks plain files?

Koncpa replied:
Yes, It was tested.

zpytela replied:
For plain files the `--` spec can be used.

zpytela commented on policy/modules/contrib/frr.if:
```diff
@@ -0,0 +1,160 @@
+
+	corecmd_search_bin($1)
+	can_exec($1, frr_exec_t)
+')
```
Blank line separator missing.

Koncpa replied:
Fixed.

zpytela commented on policy/modules/contrib/frr.if:
```diff
@@ -0,0 +1,160 @@
+	files_search_pids($1)
+	read_files_pattern($1, frr_var_run_t, frr_var_run_t)
+')
+
```
Double line separator.

Koncpa replied:
Fixed.

zpytela commented on policy/modules/contrib/frr.if:
```diff
@@ -0,0 +1,160 @@
+##	<summary>
+##	Role allowed access.
+##	</summary>
+## </param>
```
$2 is never used in the interface. Are there any plans for using it?

Koncpa replied:
It's mistake of sepolicy generate. Now It should be fixed.

zpytela commented on policy/modules/contrib/frr.te:
```diff
@@ -0,0 +1,108 @@
+files_config_file(frr_conf_t)
+
+type frr_unit_file_t;
+systemd_unit_file(frr_unit_file_t)
```
The private unit file type is not used anywhere else, is it supposed to happen so?

Koncpa replied:
@milosmalik  wrote about acceptance criteria of frr in bugzilla ticket, so I added this. And also when I look to /usr/lib/systemd/system/ is here frr.service file which is labeled as frr_unit_file_t.  But when was tested frr policy I didn find any AVC relating with frr_unit_file_t. So what do you think? Can I add some permission for frr_unit_file_t?

zpytela replied:
What I had in mind was: By default the unit file gets the shared `systemd_unit_file_t` type. It is correct as long it is not e. g. required for the application to manage the service, then it is the time for a confined type. I haven't seen such a rule in the policy, so I think it was not necessary; not incorrect though, it may be required later.

zpytela commented on policy/modules/contrib/frr.te:
```diff
@@ -0,0 +1,108 @@
+#
+# frr local policy
+#
+allow frr_t self:capability { dac_override dac_read_search chown kill net_bind_service net_raw setgid setuid };
```
I cannot find justification for dac_override/dac_read_search in this PR. Move chown in front of dac_.

Koncpa replied:
About dac_override I wrote in Ondra review, and about dac_read_search the vtysh need boot startup configuration with specific daemon, so I suppose that vtysh need read frr.conf .

Here I attach a example of full audit message about dac_read_search :
`type=PROCTITLE msg=audit(01/29/2021 10:00:36.033:3841) : proctitle=/usr/bin/vtysh -b -d zebra·
type=PATH msg=audit(01/29/2021 10:00:36.033:3841) : item=0 name=/etc/frr/frr.conf inode=262500 dev=fc:01 mode=file,640 ouid=frr ogid=frr rdev=00:00 obj=unconfined_u:object_r:frr_conf_t:s0 nametype=NORMAL cap_fp=none cap_fi=none cap_fe=0 cap_fver=0 cap_frootid=0·
type=CWD msg=audit(01/29/2021 10:00:36.033:3841) : cwd=/·
type=SYSCALL msg=audit(01/29/2021 10:00:36.033:3841) : arch=x86_64 syscall=openat success=yes exit=9 a0=0xffffff9c a1=0x5587f0b20780 a2=O_RDONLY a3=0x0 items=1 ppid=81191 pid=81243 auid=unset uid=root gid=root euid=root suid=root fsuid=root egid=root sgid=root fsgid=root tty=(none) ses=unset comm=vtysh exe=/usr/bin/vtysh subj=system_u:system_r:frr_t:s0 key=(null)·
type=AVC msg=audit(01/29/2021 10:00:36.033:3841) : avc:  denied  { dac_read_search } for  pid=81243 comm=vtysh capability=dac_read_search  scontext=system_u:system_r:frr_t:s0 tcontext=system_u:system_r:frr_t:s0 tclass=capability permissive=0·
type=AVC msg=audit(01/29/2021 10:00:36.033:3841) : avc:  denied  { dac_read_search } for  pid=81243 comm=vtysh capability=dac_read_search  scontext=system_u:system_r:frr_t:s0 tcontext=system_u:system_r:frr_t:s0 tclass=capability permissive=0`

zpytela commented on policy/modules/contrib/frr.te:
```diff
@@ -0,0 +1,108 @@
+
+domain_use_interactive_fds(frr_t)
+
+logging_send_syslog_msg(frr_t)
```
logging is not a part of base module, needs optional block. This comment also applies to usage in interfaces.

Koncpa replied:
Fixed.

zpytela commented on policy/modules/contrib/frr.te:
```diff
@@ -0,0 +1,108 @@
+
+fs_read_nsfs_files(frr_t)
+
+auth_use_nsswitch(frr_t)
```
First block of interfaces calls is for kernel, then it should be sorted alphabetically.

Koncpa replied:
Fixed.

zpytela commented on policy/modules/contrib/frr.te:
```diff
@@ -0,0 +1,108 @@
+corenet_tcp_bind_versa_tek_port(frr_t)
+corenet_udp_bind_bfd_control_port(frr_t)
+corenet_udp_bind_router_port(frr_t)
+corenet_udp_bind_cmadmin_port(frr_t)
```
Does this order follow some rule like port numbers? I'd rather see them alphabetically sorted.

Koncpa replied:
Fixed.

zpytela commented on policy/modules/contrib/frr.te:
```diff
@@ -0,0 +1,108 @@
+	modutils_read_module_config(frr_t)
+	modutils_exec_kmod(frr_t)
+	modutils_getattr_module_deps(frr_t)
+	modutils_read_module_deps_files(frr_t)
```
Can this list also be sorted?

Koncpa replied:
Fixed.

zpytela commented on policy/modules/contrib/frr.te:
```diff
@@ -0,0 +1,108 @@
+allow frr_t self:tcp_socket { connect connected_stream_socket_perms };
+allow frr_t self:udp_socket create_socket_perms;
+allow frr_t self:rawip_socket create_socket_perms;
+allow frr_t self:unix_stream_socket connectto;
```
Try to sort alphabetically whenever possible.

Koncpa replied:
Fixed.

zpytela commented on policy/modules/contrib/frr.te:
```diff
@@ -0,0 +1,108 @@
+files_pid_filetrans(frr_t, frr_var_run_t, { dir file lnk_file })
+
+allow frr_t frr_exec_t:dir search_dir_perms;
+read_lnk_files_pattern(frr_t, frr_exec_t, frr_exec_t)
```
Is there an example for executable symlink?

Koncpa replied:
Yes, here you have example. 

Example : `type=PROCTITLE msg=audit(02/02/2021 06:02:14.242:651) : proctitle=/usr/lib/frr/fabricd -d -A 127.0.0.1 
type=PATH msg=audit(02/02/2021 06:02:14.242:651) : item=0 name=/usr/lib64/frr/libfrr.so.0 nametype=UNKNOWN cap_fp=none cap_fi=none cap_fe=0 cap_fver=0 cap_frootid=0 
type=CWD msg=audit(02/02/2021 06:02:14.242:651) : cwd=/etc/frr 
type=SYSCALL msg=audit(02/02/2021 06:02:14.242:651) : arch=x86_64 syscall=openat success=no exit=EACCES(Permission denied) a0=0xffffff9c a1=0x7ffcae5eadb0 a2=O_RDONLY|O_CLOEXEC a3=0x0 items=1 ppid=35891 pid=35892 auid=unset uid=root gid=root euid=root suid=root fsuid=root egid=root sgid=root fsgid=root tty=(none) ses=unset comm=fabricd exe=/usr/lib/frr/fabricd subj=system_u:system_r:frr_t:s0 key=(null) 
type=AVC msg=audit(02/02/2021 06:02:14.242:651) : avc:  denied  { read } for  pid=35892 comm=fabricd name=libfrr.so.0 dev="vda1" ino=35651802 scontext=system_u:system_r:frr_t:s0 tcontext=system_u:object_r:frr_exec_t:s0 tclass=lnk_file permissive=0 
`

zpytela replied:
Thanks, I managed to find the previous discussion.

zpytela replied:
Can you post the output of this command?
```
ls -laZ /usr/lib64/frr
```

WOnder93 replied:
This seems like a packaging bug in `frr` - the executable files currently in `/usr/lib/frr` should be placed in `/usr/libexec/frr` instead. Did you report it to the maintainer/upstream?

Koncpa replied:
Here:

- drwxr-xr-x.  3 root root system_u:object_r:frr_exec_t:s0     113 Feb  2 04:42 .
- dr-xr-xr-x. 49 root root system_u:object_r:lib_t:s0        24576 Feb  2 04:42 ..
- lrwxrwxrwx.  1 root root system_u:object_r:frr_exec_t:s0      19 Sep  3  2019 libfrrsnmp.so.0 -> libfrrsnmp.so.0.0.0
- -rwxr-xr-x.  1 root root system_u:object_r:frr_exec_t:s0   20792 Sep  3  2019 libfrrsnmp.so.0.0.0
- lrwxrwxrwx.  1 root root system_u:object_r:frr_exec_t:s0      15 Sep  3  2019 libfrr.so.0 -> libfrr.so.0.0.0
- -rwxr-xr-x.  1 root root system_u:object_r:frr_exec_t:s0 1200856 Sep  3  2019 libfrr.so.0.0.0
- drwxr-xr-x.  2 root root system_u:object_r:frr_exec_t:s0     151 Feb  2 04:42 modules

zpytela replied:
This does not conform to the fc file. What do restorecon/matchpathcon say?
```
restorecon -Rvn /usr/lib64/frr
matchpathcon /usr/lib64/frr/libfrr.so.0 
```

WOnder93 replied:
The problem is that in .fc there is `/usr/lib/frr(/.*)?              gen_context(system_u:object_r:frr_exec_t,s0)`, which breaks the labels in `/usr/lib64/frr/...`, since there is an [equivalency](https://src.fedoraproject.org/rpms/selinux-policy/blob/master/f/file_contexts.subs_dist#_8) between `/usr/lib` and `/usr/lib64`.

Koncpa replied:
@zpytela , matchpathcon say:
`/usr/lib64/frr/libfrr.so.0	system_u:object_r:frr_exec_t:s0`

Koncpa replied:
> This seems like a packaging bug in `frr` - the executable files currently in `/usr/lib/frr` should be placed in `/usr/libexec/frr` instead. Did you report it to the maintainer/upstream?

No didn't report that.

Koncpa replied:
> The problem is that in .fc there is `/usr/lib/frr(/.*)? gen_context(system_u:object_r:frr_exec_t,s0)`, which breaks the labels in `/usr/lib64/frr/...`, since there is an [equivalency](https://src.fedoraproject.org/rpms/selinux-policy/blob/master/f/file_contexts.subs_dist#_8) between `/usr/lib` and `/usr/lib64`.

So issue is that this files should be in `/usr/libexec/frr` a also have different label?

zpytela replied:
Executable files which are not supposed to be executed from cmdline should be in /usr/libexec. Shared libraries should be in /usr/lib64 and have the common type lib_t unless there was a reason for having a private type - this matches the purpose of shared libraries.

Koncpa replied:
@WOnder93 So I have already reported this bug to frr maintainer.

zpytela commented:
@koncpa, do you want to keep the 2 commit separate, like to distinguish between the initial policy and later enhancements, or will it make sense for you to squash them?

Koncpa commented:
> @Koncpa, do you want to keep the 2 commit separate, like to distinguish between the initial policy and later enhancements, or will it make sense for you to squash them?

What do you think that will be better? I made two commits because on previous commit made @WOnder93 review so I want distinguish later enchancements for next review.

Koncpa committed:
Run frr service as frr_t and add new bfd_multi port

Confined frr service as frr_t and create policy files and rules for frr
service.
FRRouting (FRR) is an IP routing protocol which
includes protocol daemons for BGP, IS-IS, LDP, OSPF, PIM, and RIP.
Add new port called bfd_multi for bfdd daemon, which is use in
frr service.

Koncpa commented:
Squash two commits into one.

vmojzis commented:
@Koncpa Could you please rebase the PR to incoporate https://github.com/fedora-selinux/selinux-policy/pull/556/commits/abf7fa21fa5bb42e773f1255cd724745b05e79d3 ?

WOnder93 commented on policy/modules/contrib/frr.if:
```diff
@@ -0,0 +1,161 @@
+##	<summary>
+##	Domain allowed access.
+##	</summary>
+#
```
The `<param>` XML tag is left unclosed here, which is breaking the build.

Koncpa replied:
Thanks Ondřej. :)

Koncpa commented:
> @Koncpa Could you please rebase the PR to incoporate [abf7fa2](https://github.com/fedora-selinux/selinux-policy/commit/abf7fa21fa5bb42e773f1255cd724745b05e79d3) ?

Can I rebase PR to incorporate commit in this way?

zpytela commented:
Now it generally looks good.
